File: database/database-impl/src/main/java/io/ehdev/conrad/database/config/transaction/SpringTransactionProvider.java
package io.ehdev.conrad.database.config.transaction;

import org.jooq.TransactionContext;
import org.jooq.TransactionProvider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.jdbc.datasource.DataSourceTransactionManager;
import org.springframework.transaction.TransactionDefinition;
import org.springframework.transaction.TransactionStatus;
import org.springframework.transaction.support.DefaultTransactionDefinition;

public class SpringTransactionProvider implements TransactionProvider {

    private static final Logger logger = LoggerFactory.getLogger(SpringTransactionProvider.class);
    private final DataSourceTransactionManager txMgr;

    public SpringTransactionProvider(DataSourceTransactionManager txMgr) {
        this.txMgr = txMgr;
    }

    @Override
    public void begin(TransactionContext ctx) {
        logger.debug("Begin transaction");

        // This TransactionProvider behaves like jOOQ's DefaultTransactionProvider,
        // which supports nested transactions using Savepoints
        TransactionStatus tx = txMgr.getTransaction(new DefaultTransactionDefinition(TransactionDefinition.PROPAGATION_NESTED));
        ctx.transaction(new SpringTransaction(tx));
    }

    @Override
    public void commit(TransactionContext ctx) {
        logger.debug("commit transaction");

        txMgr.commit(((SpringTransaction) ctx.transaction()).tx);
    }

    @Override
    public void rollback(TransactionContext ctx) {
        logger.debug("rollback transaction");

        txMgr.rollback(((SpringTransaction) ctx.transaction()).tx);
    }
}

File: internal/shared-config/src/main/kotlin/tech/crom/config/SpringTransactionProvider.kt
package tech.crom.config

import org.jooq.Transaction
import org.jooq.TransactionContext
import org.slf4j.LoggerFactory
import org.springframework.jdbc.datasource.DataSourceTransactionManager
import org.springframework.transaction.TransactionDefinition
import org.springframework.transaction.TransactionStatus
import org.springframework.transaction.support.DefaultTransactionDefinition

internal class SpringTransactionProvider(private val txMgr: DataSourceTransactionManager) : org.jooq.TransactionProvider {

    override fun begin(ctx: TransactionContext) {
        logger.debug("Begin transaction")

        // This TransactionProvider behaves like jOOQ's DefaultTransactionProvider,
        // which supports nested transactions using Savepoints
        val tx = txMgr.getTransaction(DefaultTransactionDefinition(TransactionDefinition.PROPAGATION_NESTED))
        ctx.transaction(SpringTransaction(tx))
    }

    override fun commit(ctx: TransactionContext) {
        logger.debug("commit transaction")

        txMgr.commit((ctx.transaction() as SpringTransaction).tx)
    }

    override fun rollback(ctx: TransactionContext) {
        logger.debug("rollback transaction")

        txMgr.rollback((ctx.transaction() as SpringTransaction).tx)
    }

    internal inner class SpringTransaction(val tx: TransactionStatus) : Transaction

    companion object {
        private val logger = LoggerFactory.getLogger(SpringTransactionProvider::class.java)
    }
}
